alica_engine/src/engine/AlicaContext.cpp

```diff
@@ -18,12 +18,16 @@ constexpr uint32_t ALICA_CTX_BAD = 0xdeaddead;
 constexpr int ALICA_LOOP_TIME_ESTIMATE = 33; // ms
 
 AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
-        : _validTag(ALICA_CTX_GOOD)
+        : /* set the default logger using the comma operator before anything else. Note: the vebosity is set to INFO initially because the config file has not
+             been parsed yet. The logger is initialized here so that logs during construction are reported, specifically the logs while reading the config */
+        _alicaContextParams((AlicaLogger::set<AlicaDefaultLogger>(Verbosity::INFO, alicaContextParams.agentName), alicaContextParams))
+        , _localAgentName(alicaContextParams.agentName)
         , _configRootNode(initConfig(alicaContextParams.configPaths, alicaContextParams.agentName))
-        , _alicaContextParams(alicaContextParams)
+        , _validTag(ALICA_CTX_GOOD)
         , _clock(std::make_unique<AlicaClock>())
-        , _localAgentName(alicaContextParams.agentName)
 {
+    // reset the logger but this time the config is read
+    setLogger<AlicaDefaultLogger>();
 }
```

Comment:
I would create the following structure:

    AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
            : AlicaContext()
            , _validTag(ALICA_CTX_GOOD)
            , _alicaContextParams(alicaContextParams)
            , _localAgentName(alicaContextParams.agentName)
            , _configRootNode(initConfig(alicaContextParams.configPaths, alicaContextParams.agentName))
            , _alicaContextParams(alicaContextParams)
            , _validTag(ALICA_CTX_GOOD)
            , _clock(std::make_unique<AlicaClock>())
            , _localAgentName(alicaContextParams.agentName)
    {
        // sets logger based on read configs
        setLogger<AlicaDefaultLogger>();
    }

    AlicaContext::AlicaContext() // private ctor
    {
         // sets INFO logger, to be used by other ctors
        AlicaLogger::set<AlicaDefaultLogger>(Verbosity::INFO, alicaContextParams.agentName)
    }

feels cleaner

Reply:
if only `initConfig` needs to log stuff, perhaps don't use `Logging` there, just standard cout/cerr? Or the call itself can check whether Logger exists, and set a default one otherwise (since it seems to be related to Logging configs, makes sense that it overrides them if they don't exist yet)

Reply:
Right now it is only `initConfig()` which can use the `Logger`, but I think it's better to have the `Logger` initialized first because anyone could add code that does logging in either one of the constructors of some member variable or in the initializer list itself & then it would crash. `initConfig` already has this bug & its painstaking to look through the code to see if nothing else is invoking the logger. Therefore, I think the first option you suggested is better although I'm not familiar with its syntax, let me explore more

Reply:
looks like the first approach is not possible because the delegating constructor should be the only one in the initializer list: see https://en.cppreference.com/w/cpp/language/constructor

I also thought about outputting to cout/cerr but I guess that could be problematic if we are relying on custom logging like ROS in which case the logs to cout/cerr may not appear, therefore I decided its better to use the Logger always

Reply:
What if we delay `_configRootNode` into ctor's body?

Reply:
still leaves the problem where other constructors/initializers can end up using the logger in future.

Reply:
How about this solution:
```
AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
                : AlicaContext((AlicaLogger::set<AlicaDefaultLogger>(Verbosity::INFO, alicaContextParams.agentName), true), alicaContextParams) {}

AlicaContext::AlicaContext(bool, const AlicaContextParams& alicaContextParams) : ...
```

Reply:
@Dale-Koenig since you approved, I'm assuming you are ok with the current state here?

Reply:
because @david-simoes-93 & me could reach a conclusion, so we decided to leave it up to other reviewers

Reply:
Yeah no strong opinion